versioned_tree.go

```diff
@@ -154,34 +154,29 @@ func (tree *VersionedTree) GetVersioned(key []byte, version int64) (
 }
 
 // SaveVersion saves a new tree version to disk, based on the current state of
-// the tree. Multiple calls to SaveVersion with the same version are not allowed.
-func (tree *VersionedTree) SaveVersion(version int64) ([]byte, error) {
+// the tree. Returns the hash and new version number.
+func (tree *VersionedTree) SaveVersion() ([]byte, int64, error) {
+	version := tree.latestVersion + 1
```

Comment:
yes, you do set the version here. very nicely actually (disregard that part of my previous comment).
however, I don't like the assumption that it is increaded by clone.
rather you set the version in clone()

then it is clear what happens, eg:

* init: VersionedTree.latestVersion=0, orphanedTree.version=1
* SaveVersion:
    * v := VersionedTree.latestVersion+1
    * VersionedTreeversions[v] = ...
    * orphanTree = ...clone()
    * orphanTree.version = v+1

That is super clear and only happens in init and save, not hidden in clone, which is not expected to have side-effects.

Reply:
Yeap. Let me know what you think of the changes. We have to set the version before Save() for orphanTree because it uses it to manage orphans, but the rest is as you describe.